cocos2d/core/renderer/RendererCanvas.js

```diff
@@ -37,24 +37,166 @@ cc.rendererCanvas = {
+            var oldRegion = cmd._oldRegion;
+            var currentRegion = cmd._currentRegion;
+            if (regionFlag > cc.Node.CanvasRenderCmd.RegionStatus.NotDirty) {
+                ++dirtryRegionCount;
+                if(dirtryRegionCount > this._dirtyRegionCountThreshold)
+                    result = false;
+                //add
+                if(result) {
+                    (!currentRegion.isEmpty()) && dirtyRegion.addRegion(currentRegion);
+                    if (cmd._regionFlag > cc.Node.CanvasRenderCmd.RegionStatus.Dirty) {
+                        (!oldRegion.isEmpty()) && dirtyRegion.addRegion(oldRegion);
+                    }
+                }
+                cmd._regionFlag = cc.Node.CanvasRenderCmd.RegionStatus.NotDirty;
+            }
+
+        }
+
+        return result;
+    },
+
+    _beginDrawDirtyRegion: function (ctxWrapper) {
+        var ctx = ctxWrapper.getContext();
+        var dirtyList = this._dirtyRegion.getDirtyRegions();
+        ctx.save();
+        //add clip
+        var scaleX = ctxWrapper._scaleX;
+        var scaleY = ctxWrapper._scaleY;
+        ctxWrapper.setTransform({a: 1, b: 0, c: 0, d: 1, tx: 0, ty: 0}, scaleX, scaleY);
+        ctx.beginPath();
+        for (var index = 0, count = dirtyList.length; index < count; ++index) {
+            var region = dirtyList[index];
+            ctx.rect(region._minX , -region._maxY , region._width , region._height );
+        }
+        ctx.clip();
+        //end add clip
+    },
+
+    _endDrawDirtyRegion: function (ctx) {
+        ctx.restore();
+    },
+
+    _debugDrawDirtyRegion: function (ctxWrapper) {
+        if (!this._debugDirtyRegion) return;
+        var ctx = ctxWrapper.getContext();
+        var dirtyList = this._dirtyRegion.getDirtyRegions();
+        //add clip
+        var scaleX = ctxWrapper._scaleX;
+        var scaleY = ctxWrapper._scaleY;
+        ctxWrapper.setTransform({a: 1, b: 0, c: 0, d: 1, tx: 0, ty: 0}, scaleX, scaleY);
+        ctx.beginPath();
+        for (var index = 0, count = dirtyList.length; index < count; ++index) {
+            var region = dirtyList[index];
+            ctx.rect(region._minX, -region._maxY , region._width , region._height );
+        }
+        var oldstyle = ctx.fillStyle;
+        ctx.fillStyle = 'green';
+        ctx.fill();
+        ctx.fillStyle = oldstyle;
+        //end add clip
+    },
     /**
      * drawing all renderer command to context (default is cc._renderContext)
      * @param {cc.CanvasContextWrapper} [ctx=cc._renderContext]
      */
-    rendering: function (ctx) {
-        var locCmds = this._renderCmds, i, len,
-            scaleX = cc.view.getScaleX(),
+    rendering: function (ctxWrapper) {
+        var dirtyRegion = this._dirtyRegion = this._dirtyRegion || new cc.DirtyRegion();
+        var viewport = cc._canvas;
+        var wrapper = ctxWrapper || cc._renderContext;
+        var ctx = wrapper.getContext();
+
+        var scaleX = cc.view.getScaleX(),
```

Comment:
ok, I see, you use wrapper's scaleX & scaleY instead of calculate scale in render command, but it would affect all render commands, are you sure everything is correct with the modifications